File: 프로그래머스/3/43164. 여행경로/여행경로.py
from collections import defaultdict

def solution(tickets):
    graph = defaultdict(list)
    
    # 그래프 구성 + 정렬
    for start, end in tickets:
        graph[start].append(end)
    
    for key in graph:
        graph[key].sort(reverse = True) # 나중에 pop() 쓰려고 reverse
    
    route = []
    
    def dfs(curr):
        
        while graph[curr]:
            next = graph[curr].pop()
            #print(curr, next)
            dfs(next)
        route.append(curr) # 더이상 진행할 곳이 없으면 넣기
        #print("넣음 - "+ curr)
    
    dfs("ICN")
    
    return route[::-1] # 역순으로 뒤집기

File: 프로그래머스/3/43164. 여행경로/여행경로.kt
import java.util.PriorityQueue

class Solution {
    fun solution(tickets: Array<Array<String>>): Array<String> {
        val graph = HashMap<String, PriorityQueue<String>>()

        // 우선순위 큐를 사용해 알파벳 순으로 탐색 (사전순 정렬)
        for ((from, to) in tickets) {
            graph.computeIfAbsent(from) { PriorityQueue() }.add(to)
        }

        val route = mutableListOf<String>()

        fun dfs(airport: String) {
            val destinations = graph[airport]

            while (destinations != null && destinations.isNotEmpty()) {
                val next = destinations.poll()
                dfs(next)
            }

            // 역순으로 삽입하여 최종 경로를 구성
            route.add(airport)
        }

        dfs("ICN")

        return route.reversed().toTypedArray()
    }
}
